client/landing/stepper/declarative-flow/internals/steps-repository/site-options/index.tsx

```diff
@@ -1,9 +1,178 @@
+	const { currentSiteTitle, currentTagline } = useSelect(
+		( select ) => ( {
+			currentSiteTitle: ( select( ONBOARD_STORE ) as OnboardSelect ).getSelectedSiteTitle(),
+			currentTagline: ( select( ONBOARD_STORE ) as OnboardSelect ).getSelectedSiteDescription(),
+		} ),
+		[]
+	);
+
+	const { goBack, goNext, submit } = navigation;
+	const [ siteTitle, setSiteTitle ] = React.useState( currentSiteTitle ?? '' );
+	const [ tagline, setTagline ] = React.useState( currentTagline ?? '' );
+	const [ formTouched, setFormTouched ] = React.useState( false );
+	const intent = useSelect(
+		( select ) => ( select( ONBOARD_STORE ) as OnboardSelect ).getIntent(),
+		[]
+	);
+	const translate = useTranslate();
+	const site = useSite();
+
+	const { saveSiteSettings } = useDispatch( SITE_STORE );
+
+	const isFormDisabled = ! site;
+
+	useEffect( () => {
+		if ( ! site ) {
+			return;
+		}
+
+		if ( formTouched ) {
+			return;
+		}
+
+		setSiteTitle( site.name || '' );
+		setTagline( site.description );
+	}, [ site, formTouched ] );
+
+	const handleSubmit = async ( event: React.FormEvent ) => {
+		event.preventDefault();
+
+		if ( isFormDisabled ) {
+			return;
+		}
+
+		await saveSiteSettings( site.ID, {
+			blogname: siteTitle,
+			blogdescription: tagline,
+		} );
+
+		recordTracksEvent( 'calypso_signup_site_options_submit', {
+			has_site_title: !! siteTitle,
+			has_tagline: !! tagline,
+		} );
+
+		submit?.( { siteTitle, tagline } );
+	};
+
+	const onChange = ( event: React.FormEvent< HTMLInputElement > ) => {
+		if ( isFormDisabled ) {
+			return;
+		}
+
+		setFormTouched( true );
+
+		switch ( event.currentTarget.name ) {
+			case 'siteTitle':
+				return setSiteTitle( event.currentTarget.value );
+			case 'tagline':
+				return setTagline( event.currentTarget.value );
+		}
+	};
+
+	const getTextsFromIntent = ( intent: string ) => {
+		switch ( intent ) {
+			case 'sell':
+				return {
+					headerText: translate( "First, let's give your store a name" ),
+					headerImage: storeImageUrl,
+					siteTitleLabel: translate( 'Store name' ),
+					taglineLabel: translate( 'Tagline' ),
+					taglineExplanation: translate( 'In a few words, explain what your store is about.' ),
+				};
+			case 'write':
+				return {
+					headerText: translate( "First, let's give your blog a name" ),
+					headerImage: siteOptionsUrl,
+					siteTitleLabel: translate( 'Blog name' ),
+					taglineLabel: translate( 'Tagline' ),
+					taglineExplanation: translate( 'In a few words, explain what your blog is about.' ),
+				};
+			case 'build':
+			default:
+				return {
+					headerText: translate( "Let's give your site a name" ),
+					headerImage: siteOptionsUrl,
+					siteTitleLabel: translate( 'Site name' ),
+					taglineLabel: translate( 'Tagline' ),
+					taglineExplanation: translate( 'In a few words, explain what your site is about.' ),
+				};
+		}
+	};
+
+	const { headerText, headerImage, siteTitleLabel, taglineLabel, taglineExplanation } =
+		getTextsFromIntent( intent );
```

Comment:
Should we memo this?

Reply:
I don't think the function needs a performance improvement.